content/blog/2019-10-18-double-checked-locking/index.md

```diff
@@ -0,0 +1,407 @@
+or running in parallel on different processors simultaneously (e.g. thread 3 and thread 4),
+multiple copies of `helper` could be created.
+
+### Always-synchronized solution is slow
+
+To fix this issue, we can simply add a lock to this critical section as following,
+   so that only one thread can enter this critical section at a time.
+
+```Java
+class Foo {
+    private Helper helper = null;
+    public Helper getHelper() {
+        synchronized(this) {
+            if (helper == null)
+                helper = new Helper();
+            return helper;
+        }
+    }
+}
+```
+
+<img src="lazy-init-with-lock.png" style="width: 100%;">
+
+However, we only need this section of code to be synchronized for the first time.
+After the object is created, acquiring and releasing lock is unnecessary,
+and they can have a huge performance impact.
+
+What we want is like the following:
+
+<img src="lazy-init-ideal.png" style="width: 100%;">
+
+Only the first thread will enter the synchronized section and create the object. 
+Once the `helper` is initialized, all subsequent accesses can run in parallel
+without synchronization.
+
+
+Intuitively, we may come up with the following sequence to do this job:
+
+1. Check if the object is initialized without locking.
+If it is, then return the object immediately.
+2. Acquire the lock and check again if the object is initialized.
+If another thread has grabbed the lock before, 
+   the current thread can see the object is created, and return the object.
+3. Otherwise, the current thread will create the object and return.
+
+With the guideline above, we will get a code like following:
+
+```Java
+class Foo {
+    private Helper helper = null;
+    public Helper getHelper() {
+        if (helper == null) {              // first check
+            synchronized(this) {
+                if (helper == null)         // second check
+                    helper = new Helper();
+            }
+        }
+        return helper;
+    }
+}
+```
+
+This tactic is call double-checked locking.
+
+### Double-checked locking is broken 
+
+However, this code is not guaranteed to work.
+
+`helper = new Helper()` is not an atomic operation,
+it consists multiple instructions allocating space, initializing fields of the object,
+and assigning address to `helper`. 
+
+In order to show what is really happening there,
+we expand `helper = new Helper()` with some pseudocode 
+and inline the object initialization code.
+
+```Java
+class Foo {
+    private Helper helper = null;
+    public Helper getHelper() {
+        if (helper == null) {
+            synchronized(this) {
+                if (helper == null) {
+                    ptr = allocate();
+                    ptr.field1 = initField1();
+                    ptr.field2 = initField2();
+                    helper = ptr;
+                }
+            }
+        }
+        return helper;
+    }
+}
+```
```

Comment:
Inlining the initialization method is a really good idea! This makes it clearer.